python/pyiceberg/io/pyarrow.py

```diff
@@ -721,18 +773,38 @@ def _file_to_table(
         fragment_scanner = ds.Scanner.from_fragment(
             fragment=fragment,
             schema=physical_schema,
-            filter=pyarrow_filter,
+            # This will push down the query to Arrow.
+            # But in case there are positional deletes, we have to apply them first
+            filter=pyarrow_filter if not positional_deletes else None,
             columns=[col.name for col in file_project_schema.columns],
         )
 
+        if positional_deletes:
+            # In the case of a mask, it is a bit awkward because we first
```

Comment:
This if-branch is not in its final form (yet). With positional deletes and a user-provided filter, we fetch too many records. See https://github.com/apache/arrow/issues/35301

Reply:
Also checked it in a Notebook:

![image](https://github.com/apache/iceberg/assets/1134248/2d6392b5-9c7c-43fc-86be-5181db9e1ef4)

Reply:
If I understand correctly, the problem is that we are relying on the arrow result to correspond 1-to-1 with the records in the file so that we can use position in the DataFrame as the row position in the file.

Seems like this is a big problem. If we push the filter down and it skips even one row, we will lose the ability to correctly apply the deletes. But if we need to read deletes, we don't want to read the entire file, which could mean reading whole row groups that are unnecessary.

I think we can solve this a couple of ways. What we do in Java is project a column of row positions that we carry through. If we eliminate a whole row group, we start reading the next one with its starting position determined by the number of rows in all previous groups. I don't know if Arrow supports this, but it would need to.

Another option is to project just the filter columns for the entire file. If the filter is `id > 5 and id < 10`, then project just the `id` column. Then run the filter to produce a bitmap of selected rows, and `AND` that bitmap with the position delete bitmap. That strategy can end up reading less data and being faster, but it would also rely on support in Arrow for efficiently reading a file with a bitmap selecting rows.

Reply:
> If I understand correctly, the problem is that we are relying on the arrow result to correspond 1-to-1 with the records in the file so that we can use position in the DataFrame as the row position in the file.

Yes, this is correct.

> But if we need to read deletes, we don't want to read the entire file, which could mean reading whole row groups that are unnecessary.

Based on the row group statistics, yes.

> I don't know if Arrow supports this, but it would need to.

I don't think Arrow supports this today. I think we can even implement this on the PyIceberg side, but I don't think we should. Because:

- It is internal to PyArrow
- This would pull a lot of hot code into Python, where the GIL will slow us down.

The points that you address above are correct. At the Arrow side we're looking into implementing this:  https://github.com/apache/arrow/issues/35301

The last comment was about adding an internal index column that can be used for this purpose. This way we can combine the filters, and push this down to PyArrow (and also simplify things at the PyIceberg end, I feel like all the if-else branches make the code error prone).